projects/storefrontlib/src/shared/components/popover/popover.directive.ts

```diff
@@ -72,26 +73,57 @@ export class PopoverDirective {
    * focusable element in popover container.
    */
 
+  /**
+   * Subject which emits specific type of `PopoverEvent`.
+   */
+  eventSubject: Subject<PopoverEvent> = new Subject<PopoverEvent>();
+
+  @HostListener('keydown.escape')
+  handleEscape(event: KeyboardEvent): void {
+    event?.preventDefault();
+    this.eventSubject.next(PopoverEvent.ESCAPE_KEYDOWN);
+  }
+
   @HostListener('click', ['$event'])
+  handleClick(event: MouseEvent): void {
+    event?.preventDefault();
+    if (event?.target === this.element.nativeElement && !this.isOpen) {
+      this.eventSubject.next(PopoverEvent.OPEN);
+    } else if (this.isOpen) {
+      this.eventSubject.next(PopoverEvent.CLOSE_BUTTON_CLICK);
+    }
+  }
+
   @HostListener('keydown.enter', ['$event'])
-  handleOpen(event: MouseEvent | KeyboardEvent): void {
-    if (event.target === this.element.nativeElement) this.toggle(event);
+  @HostListener('keydown.space', ['$event'])
+  handlePress(event: KeyboardEvent): void {
+    event?.preventDefault();
+    if (event?.target === this.element.nativeElement && !this.isOpen) {
+      this.eventSubject.next(PopoverEvent.OPEN_BY_KEYBOARD);
+    } else if (this.isOpen) {
+      this.eventSubject.next(PopoverEvent.CLOSE_BUTTON_KEYDOWN);
+    }
   }
 
+  @HostListener('keydown.tab')
+  @HostListener('keydown.shift.tab')
+  handleTab(): void {
```

Comment:
Wouldn't be easier/cleaner to listen all keydown events and then recognise specific key inside?

Reply:
as we agreed in discuss, I leave it like it was